File: linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
'''
0 1 2 3 4 5 6 7 8 9
                s
                f
    p

0 1
s   
f      
'''
class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:
                return True
        return False

File: linked-list-cycle.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
//  3 2 0 4 8  7
//  ^         /
//   \       /
//     -----
// if we had 2 pointers, fast would reach nil, before slow
// otherwise fast will hit slow

// 1 2 3
//     s
//     f

// 1 2 3 4
//   s
//       f

// 1 2 3 4
// 2 4 2 4

// 1 2 3 4 5 6 7 8
// 2 4 6 8 2 4 6 8

// 1 2 3 4 5 6 7 
// 2 4 6 1 3 5 7

// 1
// s
// f

// check: is fast next slow?
// start fast at head.next, and slow at head
// loop until fast gets to end (careful it may not be able to move 2ce); return false if this happens
// if fa
// 1 2 3
//   s
//       f
func hasCycle(head *ListNode) bool {
    if head == nil { return false }
    slow, fast := head, head.Next
    for ; ; {
        if fast == slow { return true }
        if fast == nil || fast.Next == nil { break }
        slow, fast = slow.Next, fast.Next.Next
    }
    return false
}